Library/Homebrew/download_strategy.rb

```diff
@@ -472,35 +467,66 @@ def stage
 # because it lets you use a private S3 bucket as a repo for internal
 # distribution.  (It will work for public buckets as well.)
 class S3DownloadStrategy < CurlDownloadStrategy
+  attr_reader :url
+
   def _fetch
-    # Put the aws gem requirement here (vs top of file) so it's only
+    curl_download s3_url, to: temporary_path
+  end
+
+  def s3_url
+    @s3_url ||= request_s3_url
+  end
+
+  def request_s3_url
+    # Put the aws-sdk-s3 gem requirement here (vs top of file) so it's only
     # a dependency of S3 users, not all Homebrew users
-    require "rubygems"
-    begin
-      require "aws-sdk-v1"
-    rescue LoadError
-      onoe "Install the aws-sdk gem into the gem repo used by brew."
-      raise
-    end
+    require_aws_s3_sdk
 
-    if @url !~ %r{^https?://([^.].*)\.s3\.amazonaws\.com/(.+)$}
-      raise "Bad S3 URL: " + @url
-    end
-    bucket = Regexp.last_match(1)
-    key = Regexp.last_match(2)
-
-    ENV["AWS_ACCESS_KEY_ID"] = ENV["HOMEBREW_AWS_ACCESS_KEY_ID"]
-    ENV["AWS_SECRET_ACCESS_KEY"] = ENV["HOMEBREW_AWS_SECRET_ACCESS_KEY"]
-
-    obj = AWS::S3.new.buckets[bucket].objects[key]
-    begin
-      s3url = obj.url_for(:get)
-    rescue AWS::Errors::MissingCredentialsError
-      ohai "AWS credentials missing, trying public URL instead."
-      s3url = obj.public_url
+    with_env(AWS_ACCESS_KEY_ID: ENV["HOMEBREW_AWS_ACCESS_KEY_ID"],
+             AWS_SECRET_ACCESS_KEY: ENV["HOMEBREW_AWS_SECRET_ACCESS_KEY"]) do
+      # use the s3 api to get a presigned URL that
+      # curl can use with a basic GET
+      begin
+        s3_object_summary.presigned_url(:get)
+      rescue Aws::Errors::MissingCredentialsError
+        ohai "AWS credentials missing, using unsigned URL."
+        url
+      end
     end
+  end
+
+  def require_aws_s3_sdk
+    require "aws-sdk-s3"
+  rescue LoadError
+    onoe "please install the aws-sdk-s3 gem into the gem repo used by brew"
+    onoe "example: gem install aws-sdk-s3 --no-rdoc --no-ri --user-install"
```

Comment:
@amancevice sorry, can you please clarify? You're suggesting to remove line 502?

Reply:
No no, just asking a question. I have been banging my head against the wall with this problem as well -- But it turns out I misunderstood what the `--user-install` flag was doing.

I was originally confused by the error message `please install the aws-sdk-s3 gem into the gem repo used by brew` because I had no idea where to find this repo.

Please disregard.

Reply:
That said, after running `gem install aws-sdk-s3 --no-rdoc --no-ri --user-install`, I still get a `LoadError` when I try to require the `aws-sdk-s3` gem.

Reply:
@amancevice No worries, actually your input is useful. I've been stumbling over this whole approach to requiring the S3 gem as it feels 'clunky' in the face of the lot of the magic that Homebrew does around isolating its environment and the bottle vs source install phases.

I've even thought maybe we could dump out the `GEM_PATH` and `GEM_HOME` values to help the user determine where that repo is.

Right now, there seem to be situations where the user needs to install the gem to the 'homebrew vendored ruby's gem repo', but there's no interface exposed by brew to do this, so instead they have to manually set their GEM_HOME / GEM_REPO paths.

Reply:
@amancevice this is a PR so it's not on the current homebrew, which means you are going to want the `aws-sdk-v1` gem,  not `aws-sdk-s3` (which is what I'm proposing to switch to with this PR).

For your case, try

`gem install aws-sdk-v1 --no-rdoc --no-ri --user-install`

and make sure it ends up in your user dir (`$HOME/.gem/`...)

Reply:
I was actually writing my own `PresignDownloadStrategy` class so that I could use the newer gem instead of the `-v1` classic:

```ruby
class PresignDownloadStrategy < CurlDownloadStrategy
  def _fetch

    # Load AWS library
    begin
      require 'aws-sdk-s3'
    rescue LoadError
      # No idea what to do here...
      onoe "please install the aws-sdk-s3 gem into the gem repo used by brew"
    end

    # Get presigned URL
    signer = Aws::S3::Presigner.new
    uri = URI.parse @url
    url = signer.presigned_url :get_object,
                               :bucket => uri.host,
                               :key => uri.path.sub(/\A\//, '')

    # Download
    curl_download url, to: temporary_path
  end
end
```

I ended up junking it and writing my own code to compute a presigned S3 URL using the standard library (info on how to do that [here](https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-query-string-auth.html)).

It works, but definitely not ideal. For one, anyone who wants to use it has to export the `HOMEBREW_AWS_ACCESS_KEY_ID` and `HOMEBREW_AWS_SECRET_ACCESS_KEY` environment variables, which having the SDK would avoid.

Reply:
Ah yeah, that's definitely not ideal because the big advantage (imo) of the S3 library is that it will do the credentials magic of looking in your ~/.aws folder, and supporting instance profile credentials present on the AWS instance metadata endpoint.

Having to export the actual keys means needing to do all that work themselves (e.g. CURLing the metadata endpoint for the instance credential keys and then setting the env vars).

Reply:
Also the presign strategy is basically exactly what the S3DownloadStrategy is already doing, although I like how you're using a presigner directly instead of relying on the object summary. I'll have to do some experimenting with that approach when I revisit this PR this weekend (I want to experiment with Mike's suggestion of installing gems in `formula_installer.rb`)

Reply:
FWIW, this seemed to work for me:

```
GEM_HOME=$(brew --repo)/Library/Homebrew/vendor/portable-ruby/current/lib/ruby/gems/2.3.0 gem install aws-sdk-s3 --no-rdoc --no-ri
```

Reply:
I ended up writing a fetch-only formula for my colleagues, as it seemed like the fetch environment is permissive-enough to allow the installation of gems into the Homebrew repo.

One coworker had a problem because somehow Homebrew was using his system Ruby so he had to `sudo gem install aws-sdk-s3` on his own, but everyone else got by with `brew fetch private/tap/aws-sdk`

```ruby
begin
  require "aws-sdk-s3"
  ohai "The aws-sdk-s3 is already installed at #{Gem::dir}"
rescue LoadError
  ohai "Installing the aws-sdk-s3 to #{Gem::dir}"
  ENV["GEM_HOME"] = ENV["GEM_PATH"] = Gem::dir
  begin
    Gem::install "aws-sdk-s3", "~> 1.8"
    require "aws-sdk-s3"
  rescue
    odie <<~EOS
      Something went wrong attempting to install the aws-sdk-s3 gem.

      Send the output of the following to someone in Engineering:

        brew config

    EOS
  end
end
```

Reply:
The solution for this is to use `install_gem_setup_path!` but outside of `download_strategy` e.g. `formula_installer.rb`. This will ensure the gem is always installed automatically and into the right location.

Reply:
I'd suggest that that be part of this PR but if you want to make a dedicated PR just for that @amancevice it would be much appreciated.

Reply:
Thanks, @MikeMcQuaid -- I didn't mean to hijack the thread, just adding color to help @eedwards-sk & others decide how to handle this stuff.